test/helpers/kubectl.go

```diff
@@ -626,41 +626,77 @@ func (kub *Kubectl) WaitKubeDNS() error {
-// pod. If the container is not ready after timeout it returns an error. The
+// WaitForKubeDNSEntry waits until the given DNS entry exists in the kube-dns
+// service. If the container is not ready after timeout it returns an error. The
 // name's format query should be `${name}.${namespace}`. If `svc.cluster.local`
-// is not present it appends to the given name and it checks  the full FQDN
-func (kub *Kubectl) WaitForKubeDNSEntry(name string) error {
+// is not present, it appends to the given name and it checks the service's FQDN.
+func (kub *Kubectl) WaitForKubeDNSEntry(serviceName, serviceNamespace string) error {
 	svcSuffix := "svc.cluster.local"
-	logger := kub.logger.WithField("dnsName", name)
+	logger := kub.logger.WithFields(logrus.Fields{"serviceName": serviceName, "serviceNamespace": serviceNamespace})
 
-	if !strings.HasSuffix(name, svcSuffix) {
-		name = fmt.Sprintf("%s.%s", name, svcSuffix)
+	serviceNameWithNamespace := fmt.Sprintf("%s.%s", serviceName, serviceNamespace)
+	if !strings.HasSuffix(serviceNameWithNamespace, svcSuffix) {
+		serviceNameWithNamespace = fmt.Sprintf("%s.%s", serviceNameWithNamespace, svcSuffix)
 	}
 	// https://bugs.launchpad.net/ubuntu/+source/bind9/+bug/854705
 	digCMD := "dig +short %s @%s | grep -v -e '^;'"
 
+	// Also provide full dig output for debugging purposes.
+	digCmdLong := "dig %s @%s"
+
 	// If it fails we want to know if it's because of connection cannot be
 	// established or DNS does not exist.
 	digCMDFallback := "dig +tcp %s @%s"
 
-	host, _, err := kub.GetServiceHostPort(KubeSystemNamespace, "kube-dns")
+	dnsClusterIP, _, err := kub.GetServiceHostPort(KubeSystemNamespace, "kube-dns")
 	if err != nil {
 		logger.WithError(err).Error("cannot get kube-dns service IP")
 		return err
 	}
 
+	doServicesMatch := false
+
 	body := func() bool {
-		res := kub.Exec(fmt.Sprintf(digCMD, name, host))
-		if !res.WasSuccessful() {
-			_ = kub.Exec(fmt.Sprintf(digCMDFallback, name, host))
+		// Due to large TTL's for service names, check if the IP for the service
+		// stored in K8s matches the IP of the service cached in DNS. These
+		// can be different, because some tests use the same service names.
+		// Wait accordingly for services to match, and for resolving the service
+		// name to resolve via DNS.
+		if !doServicesMatch {
+
+			serviceIP, _, err := kub.GetServiceHostPort(serviceNamespace, serviceName)
+			if err != nil {
+				log.WithError(err).Errorf("cannot get service IP for service %s", serviceNameWithNamespace)
+				return false
+			}
+			res := kub.Exec(fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP))
+			serviceIPFromDNS := res.SingleOut()
+			isIP := govalidator.IsIP(serviceIPFromDNS)
+			if isIP {
+				if !strings.Contains(serviceIPFromDNS, serviceIP) {
+					logger.Debugf("service IP from DNS (%s) is not the IP for the service in Kubernetes (%s)\n", serviceIPFromDNS, serviceIP)
+					return false
+				} else {
```

Comment:
If you do a return false before, you don't need to have this else.

Reply:
Refactored according to your comments elsewhere. Fixed in next upload.